nilearn/interfaces/bids/glm.py

```diff
@@ -332,88 +254,41 @@ def save_glm_to_bids(
+    )
     statistical_maps = make_stat_maps(model, contrasts, output_type="all")
     for contrast_name, contrast_maps in statistical_maps.items():
-        # Extract stat_type
-        contrast_matrix = contrasts[contrast_name]
-        # Strings and 1D arrays are assumed to be t-contrasts
-        if isinstance(contrast_matrix, str) or (contrast_matrix.ndim == 1):
-            stat_type = "t"
-        else:
-            stat_type = "F"
-
-        # Override automatic detection with explicit type if provided
-        stat_type = contrast_types.get(contrast_name, stat_type)
-
-        # Convert the contrast name to camelCase
-        contrast_name = _clean_contrast_name(contrast_name)
-
-        # Contrast-level images
-        contrast_level_mapping = {
-            "effect_size": (
-                f"{prefix}contrast-{contrast_name}_stat-effect_statmap.nii.gz"
-            ),
-            "stat": (
-                f"{prefix}contrast-{contrast_name}_stat-{stat_type}_statmap"
-                ".nii.gz"
-            ),
-            "effect_variance": (
-                f"{prefix}contrast-{contrast_name}_stat-variance_statmap"
-                ".nii.gz"
-            ),
-            "z_score": (
-                f"{prefix}contrast-{contrast_name}_stat-z_statmap.nii.gz"
-            ),
-            "p_value": (
-                f"{prefix}contrast-{contrast_name}_stat-p_statmap.nii.gz"
-            ),
-        }
-        # Rename keys
-        renamed_contrast_maps = {
-            contrast_level_mapping.get(k, k): v
-            for k, v in contrast_maps.items()
-        }
-
-        for map_name, img in renamed_contrast_maps.items():
-            img.to_filename(out_dir / map_name)
-
+        for output_type in contrast_maps:
+            img = statistical_maps[contrast_name][output_type]
+            filename = model._reporting_data["filenames"]["statistical_maps"][
+                contrast_name
+            ][output_type]
+            img.to_filename(out_dir / filename)
+
+    logger.log("Saving contrast-level statistical maps...", verbose=verbose)
     _write_model_level_statistical_maps(model, prefix, out_dir)
 
-    # Add html report
-    glm_report = model.generate_report(contrasts=contrasts, **kwargs)
+    logger.log("Generating HTML...", verbose=verbose)
+    # generate_report can just rely on the name of the files
+    # stored in the model instance.
+    # temporarily drop verbosity to avoid generate_report
+    # logging the same thing
+    model.verbose -= 1
+    glm_report = model.generate_report(**kwargs)
+    model.verbose += 1
     glm_report.save_as_html(out_dir / f"{prefix}report.html")
 
-
-def _model_level(model):
-    from nilearn.glm.first_level import FirstLevelModel
-
-    return 1 if isinstance(model, FirstLevelModel) else 2
+    return model
 
 
 def _write_model_level_statistical_maps(model, prefix, out_dir):
-    if _model_level(model) == 2:
+    if model.__str__() == "Second Level Model":
```

Comment:
Could some of these functions actually be part of the model class definition and implementation stays in the specific model instead of putting many if checks for the model type?

Reply:
yeah I can probably improve the function that generate output filenames that covers this so this one becomes simpler